_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+
+Obviously, this code is clumsy - and not just because the utility functions are crude. No, this code is just clumsy and inflexible in itself - and what's even worse, it makes one cardinal sin.
+_It mixes the responsibilities._
+
+While it is not the gravest sin of them all, as one could argue that _traversing_ the filesystem and _printing_ the filenames are very strongly related, it would be nice to do these two separately. In fact, the added flexibility might come in handy if we want to do something more than just printing - and, as we're about to see, at pretty much no added cost.
+
+So into the forest we go, preparing our tree-like data structure:
+```haskell
+data FileTree = RegularFile File | Directory File [FileTree]
+```
+Now we can write a function that grows these trees:
+```haskell
+buildTree :: File -> IO FileTree
+buildTree file = do
+  isDirectory <- isDir file
+  if isDirectory
+  then do
+    contents <- listDir file
+    subtrees <- traverse buildTree contents
+    return (Directory file subtrees)
+  else return (RegularFile file)
+```
+And we can restore the printing capability by writing another simple function:
+```haskell
+printTree :: FileTree -> String
+printTree = unlines . printTree'
+  where printTree' (RegularFile file) = ['-':(name file)]
+        printTree' (Directory file contents) = ('-':(name file)):(subtrees contents)
+        subtrees = map prepend . concatMap printTree'
+        prepend path = ' ':path
+```
+Reasonably straightforward. We could even get away with making `FileTree` derive `Show` typeclass. The main benefit, though, is that we have separated the process of traversing the filesystem from the process of listing its contents - and should we ever need a function that goes into a directory and removes everything below a certain depth, we already have half of it implemented. Moreover, if we want to recover our initial function, we just need to compose a few blocks:
+```haskell
+tree :: File -> IO ()
+tree file = do
+  dirTree <- buildTree file
+  let formatted = printTree dirTree
+  putStrLn formatted
+```
+But can we go even further?
+
+There seems to be nothing left to extract - we have separated traversal and printing, so what more could we possibly do? But there _is_ in fact a subtle thing. We are walking along the system paths as the trees unfold before us, but we do it in distinct steps: first we take a look around to gather all the possible paths; and then we walk along them.
+
+Let's try to imagine a function like this. First, we need to be able to look around:
+```haskell
+type LookAround a = a -> [a]
+```
+Then, we need to be able to step along the way and construct subtrees - but this is simple recursion. We finally need to be able to construct our node (given the value to store inside the node and the paths to follow):
+```haskell
+type BuildNode t a = a -> [t a] -> t a
```

Comment:
Btw looking at `->`... does Haskell have an equivalent of tuples? or just `data` with named fields

Reply:
Also... IIUC this is a curried function, so in Scala terms that would probably be sth like:
```scala
type BuildNode[T[_], A] = A => List[T[A]] => T[A]
```

Are non-curried (2+)-ary functions supported by Haskell?

Reply:
In Haskell the functions are curried by default, however using them is pretty much hassle-free. For example, we can define a function:
```haskell
add3 :: Int -> Int -> Int -> Int
add3 x y z = x + y + z
```
and then we can use it however we please:
```haskell
twoPlusOnePlusThree :: Int
twoPlusOnePlusThree = add3 2 1 3

addSevenToTwoOthers :: Int -> Int -> Int
addSevenToTwoOthers = add3 7
```

We can however explicitly define a function that expects a tuple as an argument:
```haskell
sumPair :: (Int, Int) -> Int
sumPair (x, y) = x + y
```